client/asset/dcr/dcr.go

```diff
@@ -2031,27 +2039,82 @@ func (dcr *ExchangeWallet) Address() (string, error) {
 	return addr.String(), nil
 }
 
+func (dcr *ExchangeWallet) accountUnlocked(ctx context.Context, acct string) (encrypted, unlocked bool, err error) {
+	var res *walletjson.AccountUnlockedResult
+	res, err = dcr.node.AccountUnlocked(ctx, acct)
+	if err != nil {
+		err = translateRPCCancelErr(err)
+		return
+	}
+	encrypted = res.Encrypted
+	if res.Unlocked != nil { // should only be when encrypted
+		unlocked = *res.Unlocked
+	}
+	return
+}
+
 // Unlock unlocks the exchange wallet.
 func (dcr *ExchangeWallet) Unlock(pw string) error {
-	return translateRPCCancelErr(dcr.node.WalletPassphrase(dcr.ctx, pw, int64(time.Duration(math.MaxInt64)/time.Second)))
+	encryptedAcct, unlocked, err := dcr.accountUnlocked(dcr.ctx, dcr.acct)
+	if err != nil {
+		return err
+	}
+	if !encryptedAcct {
+		return translateRPCCancelErr(dcr.node.WalletPassphrase(dcr.ctx, pw, 0))
+
+	}
+	if unlocked {
+		return nil
+	}
+
+	return translateRPCCancelErr(dcr.node.UnlockAccount(dcr.ctx, dcr.acct, pw))
 }
 
 // Lock locks the exchange wallet.
 func (dcr *ExchangeWallet) Lock() error {
 	if dcr.client.Disconnected() {
 		return asset.ErrConnectionDown
 	}
+
 	// Since hung calls to Lock() may block shutdown of the consumer and thus
 	// cancellation of the ExchangeWallet subsystem's Context, dcr.ctx, give
 	// this a timeout in case the connection goes down or the RPC hangs for
 	// other reasons.
 	ctx, cancel := context.WithTimeout(dcr.ctx, 5*time.Second)
 	defer cancel()
-	return translateRPCCancelErr(dcr.node.WalletLock(ctx))
+
+	encryptedAcct, unlocked, err := dcr.accountUnlocked(ctx, dcr.acct)
+	if err != nil {
+		return err
+	}
+	if !encryptedAcct {
+		return translateRPCCancelErr(dcr.node.WalletLock(ctx))
+	}
+	if !unlocked {
+		return nil
+	}
+
+	err = dcr.node.LockAccount(dcr.ctx, dcr.acct)
+	if isAccountLockedErr(err) {
+		return nil // it's already locked
+	}
+	return translateRPCCancelErr(err)
 }
 
 // Locked will be true if the wallet is currently locked.
+// Q: why are we ignoring RPC errors in this?
 func (dcr *ExchangeWallet) Locked() bool {
+	// First return locked status of the account, falling back to walletinfo if
+	// the account is not individually password protected.
+	encrypted, unlocked, err := dcr.accountUnlocked(dcr.ctx, dcr.acct)
+	if err != nil {
+		dcr.log.Errorf("accountunlocked error: %v", err)
+		// return false // or try walletinfo???
```

Comment:
I'm not sure why we are eating any errors in this method and returning false...

Reply:
I think I just really didn't want an error return value. Would `true` be better?

Reply:
:man_shrugging: I'm inclined to keep it lest something that expects this behavior break.

Reply:
Can you mention v1.0 in the comment and maybe we'll pick it up later?